File: app/src/main/java/chuck/com/challenge/DependencyModules/NetModule.java
package chuck.com.challenge.DependencyModules;

import chuck.com.challenge.appListeners.ChuckNorrisAPI;
import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Laurence on 14/10/2016.
 */
@Module
public class NetModule {

    private String baseURL;

    public NetModule(String baseURL) {
        this.baseURL = baseURL;
    }

    @Provides
    Retrofit provideRetrofit(OkHttpClient okHttpClient) {

        return new Retrofit.Builder().baseUrl(baseURL).client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create()).build();
    }

    @Provides
    OkHttpClient provideOkHttpClient() {

        return new OkHttpClient.Builder().addInterceptor(
                new HttpLoggingInterceptor()
                        .setLevel(HttpLoggingInterceptor.Level.BODY)).build();
    }

    @Provides
    ChuckNorrisAPI provideChuckNorrisAPI(Retrofit retrofit) {
        return retrofit.create(ChuckNorrisAPI.class);
    }

}

File: app/src/main/java/chuck/com/challenge/di/modules/NetModule.kt
package chuck.com.challenge.di.modules

import chuck.com.challenge.BuildConfig
import chuck.com.challenge.NetConstants.JOKE_API_BASE_URL
import dagger.Module
import dagger.Provides
import io.reactivex.Scheduler
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Named

@Module
object NetModule {

    @JvmStatic
    @Provides
    fun provideLoggingInterceptor() = HttpLoggingInterceptor().apply {
        level = if (BuildConfig.DEBUG) {
            HttpLoggingInterceptor.Level.BODY
        } else {
            HttpLoggingInterceptor.Level.NONE
        }
    }

    @JvmStatic
    @Provides
    fun provideOkHttpClient(loggingInterceptor: HttpLoggingInterceptor): OkHttpClient =
            OkHttpClient.Builder().addNetworkInterceptor(loggingInterceptor).build()

    @JvmStatic
    @Provides
    fun provideRetrofit(okHttpClient: OkHttpClient, @Named(NAME_IO_SCHEDULER) scheduler: Scheduler): Retrofit = Retrofit.Builder()
            .baseUrl(JOKE_API_BASE_URL)
            .client(okHttpClient)
            .addConverterFactory(GsonConverterFactory.create())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(scheduler))
            .build()
}